### PR TITLE
#10-MS-WN-Next-Purchase-Date

### LINKED ISSUE #10: 10. As an item, I want my estimated next purchase date to be computed at the time my purchase is recorded in the database so the app can learn how often I buy different items.
## Summary
In order to advise users about when to purchase things, the app needs to be able to calculate that guess and store it a future   date.

The `calculateEstimate` function from the `@the-collab-lab/shopping-list-utils` module will help us know how many days in the future a purchase should happen. In order to use `calculateEstimate`, we need to know how many days have passed since the time the item was last purchased.

## Acceptance criteria
- [ ] A `getDaysBetweenDates` function is exported from `utils/dates.js`  and imported into `api/firebase.js`
  - [ ] This function takes two JavaScript Dates and return the number of days that have passed between them
- [ ] When the user purchases an item, the item’s `dateNextPurchased` property is calculated using the `calculateEstimate` function and saved to the Firestore database
  - [ ] `dateNextPurchased` is saved as **a date**, not a number

### Working with Firestore’s dates
**Firestore returns dates as special objects called `Timestamps.`** You will need to convert the Timestamp to a JavaScript Date in order to work with it and complete this feature. The `Timestamp` object has a `.toDate()` method – see [the Firestore docs](https://firebase.google.com/docs/reference/js/v8/firebase.firestore.Timestamp#todate).

### Calculating the time between JavaScript dates
Comparing and manipulating JavaScript Dates is notoriously tricky. Here's a hint: all Date objects have a `getTime()` method that returns a representation of that date in milliseconds ([MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTime)).

### Calculating the next purchase date
The `@the-collab-lab/shopping-list-utils` module should already be installed in your project. This module exports `calculateEstimate`.

`calculateEstimate` accepts three arguments. They are, in order:
- `previousEstimate` (a number): The last estimated purchase interval.
- `daysSinceLastPurchase` (a number): The number of days that have elapsed between the last purchased date and now. You will need to figure out how to calculate this argument!
- `totalPurchases` (a number): The total number of purchases made for the item,

Given this information, `calculateEstimate` returns **a whole number** indicating the number of days until the user is likely to make their next purchase.

### PR DESCRIPTION
**Pull request by wheninseattle**
## Description

This issue tackles the estimation of an item's next date of purchase, which is determined at the time of purchase. 

## Related Issue

closes #10

## Acceptance Criteria

- [x] A `getDaysBetweenDates` function is exported from `utils/dates.js`  and imported into `api/firebase.js`
  - [x] This function takes two JavaScript Dates and return the number of days that have passed between them
- [x] When the user purchases an item, the item’s `dateNextPurchased` property is calculated using the `calculateEstimate` function and saved to the Firestore database
  - [x] `dateNextPurchased` is saved as **a date**, not a number

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
| ✓   | :sparkles: New feature     |
| ✓   | :hammer: Refactoring       |

## Updates
- Added `getDaysBetweenDates` function
- Utilized `calculateEstimate` function to update `dateNextPurchased` as part of the overall `updateItem()` function.
- Added an `unCheckItem()` function to firebase. This function is for use with the function which unchecks an item if it was purchased over 1 day ago. This change allows for passing the entire item object into the `updateItem()` function to handle the business logic in `firebase.js` rather than in `listItem.jsx`. This also paves the way for us to add the _undo purchase_ functionality down the road.

### Before

N/A Not a UI feature

### After

N/A Not a UI feature

## Testing Steps / QA Criteria

1. git pull
2. git checkout MS-WN-Next-Purchase-Date
3. Open existing list (we used 'my test list' for testing)
4. [Open firebase console](https://console.firebase.google.com/project/tcl-47-smart-shopping-li-6b6a7/firestore/data/~2Fexcel%20allied%20tulane~2FjsCMh2mhXzrv67mNIeLC)
5. Ensure that when you check an item, that the `totalPurchases`, `dateLastPurchased`, and  `dateNextPurchased` fields are all updated. 
6. Try testing an item that has been purchased before and an item that is being purchased for the first time to ensure null values are being handled.

### CODE REVIEW

**Review by a-cole55 — APPROVED** [review]
This code looks great. It's so clean and easy to read. You guys did an awesome job!

**Review by tmflower — APPROVED** [review]
Great job on this issue, @mariahschock and @wheninseattle! I like how readable this is.

I noticed that if I add an item and purchase it on the same day, the dateNextPurchased will be recorded as today, not a future date. Looking ahead, when we get to sorting items by dateNextPurchased, I wonder if we will want to have this case handled differently.

**Comment by wheninseattle** [comment]
@tmflower thanks for flagging that. I spent a bit of time looking into it. Initially, I thought this was because `previousEstimate` was getting the correct arguments if `!dateLastPurchased`. When I looked closer at how `calculateEstimate()` was working I noticed that in the case of an item being added and purchased in the same day, the updated estimate was < 1.

File: src/api/firebase.js [hunk 1]
```diff
@@ -88,17 +91,87 @@ export async function addItem(listId, { itemName, daysUntilNextPurchase }) {
...
+		// days since the item was created
+		const daysSinceLastTransaction = dateLastPurchased
+			? getDaysBetweenDates(dateLastPurchased)
+			: getDaysBetweenDates(dateCreated);
+
+		const updatedPurchaseCount = totalPurchases + 1;
+
+		console.log('previousEstimate', previousEstimate);
+		console.log('daysSinceLastTransaction', daysSinceLastTransaction);
+		console.log('updatedPurchaseCount', updatedPurchaseCount);
```
Comment (veeteeran) [inline src/api/firebase.js]: Please remove console logs

File: src/api/firebase.js [hunk 2]
```diff
@@ -121,21 +121,19 @@ export async function updateItem(listId, item) {
 		totalPurchases,
 	} = item;
 
-	let currentTime = new Date();
+	let currentDate = new Date();
```
Comment (veeteeran) [inline src/api/firebase.js]: nit. Does this variable ever get reassigned? If not, should be a `const`

File: src/components/ListItem.jsx [hunk 3]
```diff
@@ -12,28 +12,12 @@ export function ListItem({ name, item, listToken }) {
 			(currentTimeInSeconds - dateLastPurchased.seconds) / 60;
 		const minutesBeforeReset = 1440; // 24 hours x 60 minutes per hour
 		if (minutesSinceLastPurchased >= minutesBeforeReset) {
-			const itemData = {
-				isChecked: false,
-			};
-			updateItem(listToken, id, itemData);
+			unCheckItem(listToken, id);
```
Comment (veeteeran) [inline src/components/ListItem.jsx:15]: Nice work keeping this file clean!

File: src/utils/dates.js [hunk 4]
```diff
@@ -10,3 +10,24 @@ const ONE_DAY_IN_MILLISECONDS = 86400000;
...
+	if (newerDate) {
+		return (
+			Math.abs(newerDate.toDate() - olderDate.toDate()) /
+			ONE_DAY_IN_MILLISECONDS
+		);
+	} else {
+		const currentDate = new Date();
+		return Math.abs(currentDate - olderDate.toDate()) / ONE_DAY_IN_MILLISECONDS;
+	}
+}
```
Comment (veeteeran) [inline src/utils/dates.js:33]: This is amazing!!!

**Review by veeteeran — APPROVED** [review]
Nice work breaking the code off into smaller functions and moving them outside of the component. Future you will be so happy you did this!

**Pull request merged**